src/main/java/gregtech/api/unification/material/Material.java

```diff
@@ -0,0 +1,848 @@
+        this.chemicalFormula = withFormatting ? SmallDigits.toSmallDownNumbers(formula) : formula;
+        return this;
+    }
+
+    public ImmutableList<MaterialStack> getMaterialComponents() {
+        return materialInfo.componentList;
+    }
+
+    private Material(@Nonnull MaterialInfo materialInfo, @Nonnull MaterialProperties properties, @Nonnull MaterialFlags flags) {
+        this.materialInfo = materialInfo;
+        this.properties = properties;
+        this.flags = flags;
+
+        this.properties.setMaterial(this);
+        registerMaterial(this);
+    }
+
+    // thou shall not call
+    protected Material(String name) {
+        materialInfo = new MaterialInfo(0, name);
+        materialInfo.iconSet = MaterialIconSet.NONE;
+        properties = new MaterialProperties();
+        flags = new MaterialFlags();
+    }
+
+    protected void registerMaterial(Material material) {
+        MaterialRegistry.register(this);
+    }
+
+    public void addFlag(MaterialFlag... flags) {
+        if (MaterialRegistry.isFrozen())
+            throw new IllegalStateException("Cannot add flag to material when registry is frozen!");
+        this.flags.addFlags(flags).verify(this);
+    }
+
+    //@ZenMethod("hasFlagRaw")
+    public boolean hasFlag(MaterialFlag flag) {
+        return flags.hasFlag(flag);
+    }
+
+    public boolean hasFlags(MaterialFlag... flags) {
+        return Arrays.stream(flags).allMatch(this::hasFlag);
+    }
+
+    //@ZenMethod
+    //public boolean hasFlag(String flagName) {
+    //    long materialFlagId = MaterialFlags.resolveFlag(flagName, getClass());
+    //    return hasFlag(materialFlagId);
+    //}
+
+    protected void calculateDecompositionType() {
+        if (!materialInfo.componentList.isEmpty() &&
+            !hasFlag(MaterialFlags.DECOMPOSITION_BY_CENTRIFUGING) &&
+            !hasFlag(MaterialFlags.DECOMPOSITION_BY_ELECTROLYZING) &&
+            !hasFlag(MaterialFlags.DISABLE_DECOMPOSITION)) {
+            boolean onlyMetalMaterials = true;
+            for (MaterialStack materialStack : materialInfo.componentList) {
+                Material material = materialStack.material;
+                onlyMetalMaterials &= material.hasProperty(PropertyKey.INGOT);
+            }
+            //allow centrifuging of alloy materials only
+            if (onlyMetalMaterials) {
+                flags.addFlags(MaterialFlags.DECOMPOSITION_BY_CENTRIFUGING);
+            } else {
+                flags.addFlags(MaterialFlags.DECOMPOSITION_BY_ELECTROLYZING);
+            }
+        }
+    }
+
+    public Fluid getFluid() {
+        FluidProperty prop = getProperty(PropertyKey.FLUID);
+        if (prop == null)
+            throw new IllegalArgumentException("Material " + materialInfo.name + " does not have a Fluid!");
+        return prop.getFluid();
+    }
+
+    public FluidStack getFluid(int amount) {
+        return new FluidStack(getFluid(), amount);
+    }
+
+    public int getHarvestLevel() {
```

Comment:
Could we get a setter for harvest level? I know people have had to write reflection addons to change the harvest level of GT ores

Reply:
All of these getters and setters are intended to be cleaned up and fleshed out once CT compatibility is added, which I plan on doing after this is merged just to not hold this PR up forever